### PR TITLE
SLING-11777

### PR DESCRIPTION
**Pull request by niekraaijmakers**
Currently we are using repoinit statements with the update / merge filter, to provide sling:Folder as default folder in sling initial content. However with filevault 3.6.0+ this approach no longer works.

New approach is to inject a .content.xml file with sling:Folder in the generated sling initial content packages where there is none present (up to /jcr:root/apps). We now use the replace / merge_properties instead. Code contains failing unit tests as well.

### CODE REVIEW

**Comment by niekraaijmakers** [comment]
@rombert @anchela @karlpauls

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 1]
```diff
@@ -106,10 +121,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+        for(String repositoryPath: collectedSlingInitialContentBundleEntries.stream()
+                .map(SlingInitialContentBundleEntryMetaData::getRepositoryPath)
+                .collect(Collectors.toList())){
+            addDefaultSlingContentXMLAndParents(repositoryPath);
+        }
+    }
+
+    private void addDefaultSlingContentXMLAndParents(String repositoryPath) throws UnsupportedEncodingException {
+
+        if(StringUtils.countMatches(repositoryPath, '/') < 2){
```
Comment (rombert) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:146]: I find this opaque - please document the reason.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 2]
```diff
@@ -190,6 +249,21 @@ private SlingInitialContentBundleEntryMetaData createSlingInitialContentBundleEn
         String repositoryPath = (target != null ? target : "/") + URLDecoder.decode(entryName.substring(pathEntryValue.getPath().length()), "UTF-8");
         return new SlingInitialContentBundleEntryMetaData(targetFile, pathEntryValue, repositoryPath);
     }
+    
+
+    private PathEntry getPathEntryFromRepositoryPath(@NotNull String repositoryPath) {
+
+        if(StringUtils.countMatches(repositoryPath, '/') < 2){
```
Comment (rombert) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:256]: I find this opaque - please document the reason. Since we use it in two places, it would be good I think to extract to a method or at least have a contant whose name explains what it does.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java [hunk 3]
```diff
@@ -70,7 +67,8 @@ class BundleSlingInitialContentJarEntryExtractor {
      */
     void extractAndAddToAssembler(@NotNull BundleSlingInitialContentExtractContext context,
                                   @NotNull SlingInitialContentBundleEntryMetaData slingInitialContentBundleEntryMetaData,
-                                  @NotNull Set<SlingInitialContentBundleEntryMetaData> collectedSlingInitialContentBundleEntries) throws IOException, ConverterException {
+                                  @NotNull Set<SlingInitialContentBundleEntryMetaData> collectedSlingInitialContentBundleEntries,
+                                  @NotNull boolean isDefaultContentXmlFile) throws IOException, ConverterException {
```
Comment (rombert) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java:71]: Does `@NotNull` make sense for a primitive value?

**Comment by rombert** [comment]
> @rombert @anchela @karlpauls

I assume you wanted us as reviewers, so I added review requests.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 4]
```diff
@@ -56,10 +62,13 @@ class SlingInitialContentBundleEntryMetaDataCollector {
     private final String basePath;
     private final ContentPackage2FeatureModelConverter contentPackage2FeatureModelConverter;
     private final Path newBundleFile;
-    private final Set<SlingInitialContentBundleEntryMetaData> collectedSlingInitialContentBundleEntries = new HashSet<>();
+    private final NavigableSet<SlingInitialContentBundleEntryMetaData> collectedSlingInitialContentBundleEntries = new ConcurrentSkipListSet<>();
+    private final NavigableSet<SlingInitialContentBundleEntryMetaData> defaultContentXmlEntries = new ConcurrentSkipListSet<>();
     private final AtomicLong total = new AtomicLong(0);
     private final JarFile jarFile;
 
+    private final File DEFAULT_NODE_TYPE_XML;
```
Comment (rombert) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:70]:
```suggestion
    private final File EMPTY_SLING_FOLDER_XML;
```

And please rename the corresponding file as well.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 5]
```diff
@@ -106,10 +121,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
                 }
             }
         }
-
+
```
Comment (rombert) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:124]:
```suggestion

```

(tried to suggest to remove the whitespace-only change

**Review by rombert — CHANGES_REQUESTED** [review]
Not really a specialist in this area, but my main question is around "hardcoding" the depth of the folders that we create the intermediate node paths for.

I assume that you have already tested the "old" and "new" FileVault versions and that the content packages are installed.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java (hunk 3 above)
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java:71]: nope it doesn't :)
sonar will complain about it.....

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/AssemblerProvider.java [hunk 6]
```diff
@@ -99,9 +99,9 @@ private VaultPackageAssembler lazyConstruct(@NotNull BundleSlingInitialContentEx
     private void addPathFilterSetToAssemblerFilter(@NotNull PathEntry pathEntry, @NotNull VaultPackageAssembler assembler) {
         ImportMode importMode;
         if (pathEntry.isOverwrite()) {
-            importMode = ImportMode.UPDATE;
+            importMode = ImportMode.REPLACE;
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/AssemblerProvider.java:102]: @niekraaijmakers , i had to lookup the definition of pathEntry.isOverwrite() which points to 
```
public abstract class ImportOptions {

	/**
	 * Specifies whether imported nodes should overwrite existing nodes.
	 * NOTE: this means the existing node will be deleted and a new node
	 * will be created in the same location.
	 * @return true to overwrite nodes, false otherwise
	 */
	public abstract boolean isOverwrite();
```
while i agree that ImportMode.REPLACE is probably the better match (see https://jackrabbit.apache.org/filevault/apidocs/org/apache/jackrabbit/vault/fs/api/ImportMode.html), i am wondering why we had UPDATE before. do you recall that?

i am bit concerned that changing it will cause regressions. in any case it would be great to have some dedicated tests for that illustrating the difference in behavior.

specially since there is probably not a 1:1 match with filevault.... e.g. there is also PathEntry.isPropertyOverwrite and isOverwrite is defined to only affect nodes.... while i believe that the fv-REPLACE affects both nodes and props, no?

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/AssemblerProvider.java [hunk 7]
```diff
@@ -99,9 +99,9 @@ private VaultPackageAssembler lazyConstruct(@NotNull BundleSlingInitialContentEx
...
         } else {
-            importMode = ImportMode.MERGE;
+            importMode = ImportMode.MERGE_PROPERTIES;
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/AssemblerProvider.java:104]: is this change absolutely required for the task at hand?
do you have a test illustrating the effect?

i am not opposed but want to avoid introducing new problems.
maybe @kwin can also provide some input to make sure we understand any potential unwanted side-effects

File: src/main/java/org/apache/sling/feature/cpconverter/ContentPackage2FeatureModelConverter.java [hunk 8]
```diff
@@ -328,7 +328,6 @@ private void secondPass(@NotNull Collection<VaultPackage> orderedContentPackages
                     // finally serialize the Feature Model(s) file(s)
 
                     aclManager.addRepoinitExtension(assemblers, featuresManager);
-                    bundleSlingInitialContentExtractor.addRepoInitExtension(assemblers, featuresManager);
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/ContentPackage2FeatureModelConverter.java:331]: hm..... correct me if i am wrong.... looking at the whole class it seems to me that you are removing the only place where the `bundleSlingInitialContentExtractor` is actually being invoked making changes to the sling initial content.
am i missing something?

or phrased differently:
what is left after removing that statement is
- line 293: bundleSlingInitialContentExtractor.reset()
- line 342: bundleSlingInitialContentExtractor.reset()
- line 528: setBundleSlingInitialContentExtractor(BundleSlingInitialContentExtractor) invoked from the launcher.

and what is the reset doing now? and why do you still need it here?

it is effectively used in `DefaultEntryHandlersManager` which passes it to `BundleEntryHandler`, no? why still referencing it in the converter?

it might all be ok and needed but in that case i believe a bit of comment explaining why you are resetting it, might help. alternatively, leave that all to the handlersmanager.

**Review by anchela — CHANGES_REQUESTED** [review]
hi @niekraaijmakers , i am not yet done and will continue with the extractor but wanted to share the first bunch of findings.

in general i makes me a bit suspicious that you make all these changes without having to touch any test-cases. it leaves me with the impression that we need some extra test coverage here illustrating and verifying the expected behavior of this sling-initial-content handling

**Comment by rombert** [comment]
> in general i makes me a bit suspicious that you make all these changes without having to touch any test-cases. it leaves me with the impression that we need some extra test coverage here illustrating and verifying the expected behavior of this sling-initial-content handling

@anchela - see https://issues.apache.org/jira/browse/SLING-11778 for some ideas I had regarding improving testing. We do quite some testing around how the generated feature models ( and content packages, bundles, repoinit, etc) look, but we don't really check that they are actually installable. Having a couple of such ITs would be great IMO.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java [hunk 9]
```diff
@@ -84,7 +82,7 @@ void extractAndAddToAssembler(@NotNull BundleSlingInitialContentExtractContext c
             VaultPackageAssembler packageAssembler = assemblerProvider.initPackageAssemblerForPath(context, repositoryPath, pathEntryValue);
 
             final ContentReader contentReader = contentReaderProvider.getContentReaderForEntry(file, pathEntryValue);
-            if (contentReader != null) {
+            if (contentReader != null && !isDefaultContentXmlFile) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java:85]: if `isDefaultContentXmlFile` is false you don't need to obtain the `contentReader` to start with, no?
i am not sure i understand why this is the case, but it would make the code more readable if you would re-arrange it and only retrieve the contentreader if needed.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java [hunk 10]
```diff
@@ -117,17 +115,21 @@ void extractAndAddToAssembler(@NotNull BundleSlingInitialContentExtractContext c
...
-            try (Archive virtualArchive = SingleFileArchive.fromPathOrInputStream(tmpDocViewInputFile, bundleFileInputStream,
-                    () -> Files.createTempFile(context.getConverter().getTempDirectory().toPath(), "initial-content", Text.getName(file.getName())), contentPackageEntryPath)) {
-                // in which content package should this end up?
-
-                if (tmpDocViewInputFile != null) {
-                    packageAssembler.addEntry(contentPackageEntryPath, tmpDocViewInputFile.toFile());
-                } else {
-                    packageAssembler.addEntry(contentPackageEntryPath, bundleFileInputStream);
+            //if we are writing in a .content.xml file as default, 
+            // but the .content.xml is already present (written by specifications, then we don't do anything.)
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java:119]: i think the ) needs to go in a different place to make it clear, no?

`// but the .content.xml is already present (written by specifications), then we don't do anything.`

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java [hunk 11]
```diff
@@ -117,17 +115,21 @@ void extractAndAddToAssembler(@NotNull BundleSlingInitialContentExtractContext c
...
+            if(!(isDefaultContentXmlFile && packageAssembler.alreadyPresent(contentPackageEntryPath))){
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentJarEntryExtractor.java:120]: nitpicking: you have a space missing after the if.

the serious part:
i am trying to map your comment above to this if-clause.... you only want to skip this step if `isDefaultContentXmlFile` is `true` and the entry is already present, right?
if `isDefaultContentXmlFile` is false you don't care about the collision and if the entry does not exist you can safely ignore the `isDefaultContentXmlFile`... is that what you are saying?

why? and how did it work before? because the check for alreadyPresent was not present before, right?
please elaborate in the comment for the next dev looking at this being able to follow you

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaData.java [hunk 12]
```diff
@@ -74,4 +74,11 @@ public String toString() {
                 "repositoryPath='" + repositoryPath + '\'' +
                 '}';
     }
+
+    @Override
+    public int compareTo(@NotNull SlingInitialContentBundleEntryMetaData o) {
+        return this.repositoryPath.compareTo(o.repositoryPath);
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaData.java:80]: i am not convinced that string comparison is sensible for paths. what is the goal here?

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 13]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+    Set<SlingInitialContentBundleEntryMetaData> getDefaultContentXmlEntries() throws UnsupportedEncodingException {
+        
+        createDefaultContentXMLFiles();
+        return defaultContentXmlEntries;
+        
+    }
+
+    private void createDefaultContentXMLFiles() throws UnsupportedEncodingException {
+        
+        for(String repositoryPath: collectedSlingInitialContentBundleEntries.stream()
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:148]: nitpicking: missing space between for and (

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 14]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+                .map(SlingInitialContentBundleEntryMetaData::getRepositoryPath)
+                .collect(Collectors.toList())){
+            addDefaultSlingContentXMLAndParents(repositoryPath);
+        }
+    }
+
+    private void addDefaultSlingContentXMLAndParents(String repositoryPath) throws UnsupportedEncodingException {
 
+        if(StringUtils.countMatches(repositoryPath, '/') < 2){
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:157]: nitpicking: missing space between if and (

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 15]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+            //we don't add the primary type of level 1 and 2.
+            return;
+        }
+        
+        String parentPath = StringUtils.substringBeforeLast(repositoryPath, "/");
+        //if we already got an entry with the path defined, we skip ahead.
+        if( !alreadyFound(collectedSlingInitialContentBundleEntries, parentPath) &&
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:164]: space missing. please reformat your code to make it consistent with the rest of the module

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 16]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+            !alreadyFound(defaultContentXmlEntries, parentPath)
+           ){
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:166]: again

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 17]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+            final PathEntry pathEntryValue = getPathEntryFromRepositoryPath(repositoryPath);
+            if(pathEntryValue != null){
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:168]: again 2 times

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java (hunk 2 above)
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:267]: and add a space after if and before { (or let the IDE fix the formatting)

File: src/main/java/org/apache/sling/feature/cpconverter/vltpkg/VaultPackageAssembler.java [hunk 18]
```diff
@@ -233,6 +233,11 @@ public void addEntry(@NotNull String path, @NotNull InputStream input) throws IO
         }
     }
 
+    public @NotNull boolean alreadyPresent(@NotNull String path){
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/vltpkg/VaultPackageAssembler.java:236]: nitpicking: missing space before {

**Review by anchela — CHANGES_REQUESTED** [review]
more questions, comments and a bit of nitpicking. it would be good if you could just fix formatting.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentExtractor.java [hunk 19]
```diff
@@ -103,11 +111,7 @@ static Version getModifiedOsgiVersion(@NotNull Version originalVersion) {
     }
 
     public void reset() {
-        parentFolderRepoInitHandler.reset();
-    }
-
-    public void addRepoInitExtension(@NotNull List<VaultPackageAssembler> assemblers, @NotNull FeaturesManager featureManager) throws IOException, ConverterException {
-        parentFolderRepoInitHandler.addRepoinitExtension(assemblers, featureManager);
+
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/BundleSlingInitialContentExtractor.java:114]: this method is empty now.... and it is not inherited from any interface or superclass.... 

IMHO we can get rid of this altogether, right?
which also means (i looked for usages) that calling this `reset()` in the `ContentPackage2FeatureModelConverter` is no longer needed, which circles back to my comment about the converter class.... if i am not mistaken you could actually get rid of having the extractor passed to the converter altogether, no?

**Review by anchela — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 20]
```diff
@@ -51,15 +59,19 @@ class SlingInitialContentBundleEntryMetaDataCollector {
     private static final double THRESHOLD_RATIO = 10;
     private static final int BUFFER = 512;
     private static final long TOOBIG = 0x6400000; // Max size of unzipped data, 100MB
+    static final ClassLoader CL = SlingInitialContentBundleEntryMetaDataCollector.class.getClassLoader();
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:62]: it's only used here. make it private.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 21]
```diff
@@ -68,6 +80,20 @@ class SlingInitialContentBundleEntryMetaDataCollector {
         this.contentPackage2FeatureModelConverter = contentPackage2FeatureModelConverter;
         this.newBundleFile = newBundleFile;
         this.jarFile = context.getJarFile();
+        
+        this.defaultContentXmlFile = new File(this.basePath, "defaultContentXml.xml");
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:84]: i don't remember having ever seen the name `defaultContentXml` in the context of content packages. is this related to sling-initial content?
if not and this is something you just need here, i would be great if you could add a comment elaborating what this is used for.

Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:84]: below you use the string 'default-content.xml'

**Review by anchela — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java (hunk 21 above)
Comment (niekraaijmakers) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:84]: We could rename it.. basically its .content.xml file with sling:Folder as jcr:primaryType. We put it in sling initial content if there is no specification.

Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:84]: and .content.xml does not work? or what is the reason?
in either case: i would recommend to then add some comment into the xml file indicating that this was injected by the converter.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 22]
```diff
@@ -68,6 +80,20 @@ class SlingInitialContentBundleEntryMetaDataCollector {
...
+        writeDefaultXMLFile();
+    }
+
+    private void writeDefaultXMLFile() {
+        try(OutputStream outputStream = new FileOutputStream(defaultContentXmlFile)){
+            IOUtils.copy(Objects.requireNonNull(CL.getResourceAsStream("default-content.xml")), outputStream);
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException("Default Content XML file not found in classpath!", e);
+        } catch (IOException e) {
+            throw new RuntimeException("Could not prepare the default content XML file for sling initial content!", e);
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:94]: same: don't throw runtime-exception

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 23]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
                 }
             }
         }
-
+        
         return collectedSlingInitialContentBundleEntries;
     }
+    
+    Set<SlingInitialContentBundleEntryMetaData> getDefaultContentXmlEntries() throws UnsupportedEncodingException {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:139]: annotate return value with @NotNull

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 24]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+    private void createDefaultContentXMLFiles() throws UnsupportedEncodingException {
+        
+        for(String repositoryPath: collectedSlingInitialContentBundleEntries.stream()
+                .map(SlingInitialContentBundleEntryMetaData::getRepositoryPath)
+                .collect(Collectors.toList())){
+            addDefaultSlingContentXMLAndParents(repositoryPath);
+        }
+    }
+
+    private void addDefaultSlingContentXMLAndParents(String repositoryPath) throws UnsupportedEncodingException {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:155]: UnsupportedEncodingException is never thrown (according to my IDE). please remove it.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 25]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
     }
+    
+    Set<SlingInitialContentBundleEntryMetaData> getDefaultContentXmlEntries() throws UnsupportedEncodingException {
+        
+        createDefaultContentXMLFiles();
+        return defaultContentXmlEntries;
+        
+    }
+
+    private void createDefaultContentXMLFiles() throws UnsupportedEncodingException {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:146]: `UnsupportedEncodingException` is never thrown if you removed the unnecessary throws clause from `addDefaultSlingContentXMLAndParents`

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java (hunk 23 above)
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:139]: `UnsupportedEncodingException` is never thrown if you remove the unnecessary clause from `createDefaultContentXMLFiles` and `addDefaultSlingContentXMLAndParents`

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 26]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+            if(pathEntryValue != null){
+                defaultContentXmlEntries.add(new SlingInitialContentBundleEntryMetaData(defaultContentXmlFile, pathEntryValue, parentPath + "/" + DOT_CONTENT_XML));
+            }
+         
+        }
+        addDefaultSlingContentXMLAndParents(parentPath);
+        
+    }
+    
+    private boolean alreadyFound(Set<SlingInitialContentBundleEntryMetaData> set, String parentPath){
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:177]: add @NotNull annotations for the params

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java (hunk 24 above)
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:155]: add @NotNull annotation to the repositoryPath param

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java (hunk 1 above)
Comment (anchela) [reply src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:157]: also: could it be that the repositorypath contains a trailing / ? then you would actually continue.....
but similar to roberts finding: what is different with level 1 and 2? that looks brittle....

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 27]
```diff
@@ -106,10 +132,54 @@ Set<SlingInitialContentBundleEntryMetaData> collectFromContextAndWriteTmpFiles()
...
+    }
+
+    private void addDefaultSlingContentXMLAndParents(String repositoryPath) throws UnsupportedEncodingException {
 
+        if(StringUtils.countMatches(repositoryPath, '/') < 2){
+            //we don't add the primary type of level 1 and 2.
+            return;
+        }
+        
+        String parentPath = StringUtils.substringBeforeLast(repositoryPath, "/");
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:162]: again... what if the path ends with trailing /?
i would strongly recommend to use utilities methods from `org.apache.jackrabbit.util.Text` instead. getRelativeParent is probably what you need here.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java (hunk 26 above)
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:177]: i believe `alreadyFound` can be static.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 28]
```diff
@@ -190,6 +260,21 @@ private SlingInitialContentBundleEntryMetaData createSlingInitialContentBundleEn
         String repositoryPath = (target != null ? target : "/") + URLDecoder.decode(entryName.substring(pathEntryValue.getPath().length()), "UTF-8");
         return new SlingInitialContentBundleEntryMetaData(targetFile, pathEntryValue, repositoryPath);
     }
+    
+
+    private PathEntry getPathEntryFromRepositoryPath(@NotNull String repositoryPath) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:265]: add @Nullable annotation for return value.

File: src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java [hunk 29]
```diff
@@ -190,6 +260,21 @@ private SlingInitialContentBundleEntryMetaData createSlingInitialContentBundleEn
...
+
+        if(StringUtils.countMatches(repositoryPath, '/') < 2){
+            //we don't add the primary type of level 1 and 2.
+            return null;
+        }
+        
+        Optional<PathEntry> pathEntryOptional = context.getPathEntryList().stream().filter(
+                pathEntry -> checkIfPathStartsWithOrIsEqual(pathEntry.getTarget(), repositoryPath)
```
Comment (anchela) [inline src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/SlingInitialContentBundleEntryMetaDataCollector.java:273]: instead of creating `checkIfPathStartsWithOrIsEqual` yourself, i would strongly recommend to use again Text utility. `isDescendantOrEqual`

**Comment by niekraaijmakers** [comment]
After internal discussions, we decided to go for the original approach of using repoinit statement (this time making sure only 1 per path gets outputted with the proper primary type) and leaving the filters untouched. Closing PR for now.

**Review by anchela — CHANGES_REQUESTED** [review]
